### PR TITLE
ensure the SkipAdminCheck attribute is set for Java21 runtime

### PR DESCRIPTION
**Pull request by lachlan-roberts**
Ensure that the SkipAdminCheck attribute is set on the request if the `x-google-internal-skipadmincheck` header is found.

This was done by the `RpcConnector` used by jetty 9.4, but was not being done by the `DelegateRpcExchange` for the jetty12 runtimes.

### CODE REVIEW

**Review by ludoch — APPROVED** [review]

**Comment by gchux** [comment]
something interesting: for `EE10` even without this change, and the request attribute `com.google.apphosting.internal.SkipAdminCheck` being `null` the request still flows all the way to the App's HTTP handler:

image

**Comment by gchux** [comment]
I personally think that `JettyServletEngineAdapter` is a better place to introduce this logic.

File: runtime/runtime_impl_jetty12/src/main/java/com/google/apphosting/runtime/jetty/delegate/impl/DelegateRpcExchange.java [hunk 1]
```diff
@@ -56,6 +62,34 @@ public DelegateRpcExchange(RuntimePb.UPRequest request, MutableUpResponse respon
     HttpMethod method =
         LEGACY_MODE ? HttpMethod.INSENSITIVE_CACHE.get(protocol) : HttpMethod.CACHE.get(protocol);
     _httpMethod = method != null ? method.asString() : protocol;
+
+    final boolean skipAdmin = hasSkipAdminCheck(request);
```
[the remark this answers is not in the document]
Comment (lachlan-roberts) [reply runtime/runtime_impl_jetty12/src/main/java/com/google/apphosting/runtime/jetty/delegate/impl/DelegateRpcExchange.java:66]: This should be used in both the RPC path for java8 and the HTTP->RPC path in the GEN2 Jetty9.4 runtimes.

**Comment by lachlan-roberts** [comment]
> I personally think that `JettyServletEngineAdapter` is a better place to introduce this logic.

`new DelegateRpcExchange(upRequest, upResponse)` is called directly by `JettyServletEngineAdapter.serviceRequest`. And its not just about setting the attribute, we also want to mark the `DelegateRpcExchange` as secure even though it might have been `http` not `https` so it bypasses the CONFIDENTIAL transport guarantee.

So I think it should be done in the `DelegateRpcExchange` like this.

**Comment by lachlan-roberts** [comment]
> something interesting: for EE10 even without this change, and the request attribute com.google.apphosting.internal.SkipAdminCheck being null the request still flows all the way to the App's HTTP handler:

looking into this now

**Comment by gregw** [comment]
> something interesting: for `EE10` even without this change, and the request attribute `com.google.apphosting.internal.SkipAdminCheck` being `null` the request still flows all the way to the App's HTTP handler:

@gchux Servlet constraints combine in an non obvious way.  The weaker of the constraints typically wins in the merge. See https://jakarta.ee/specifications/servlet/6.0/jakarta-servlet-spec-6.0#combining-constraints

**Review by ludoch — APPROVED** [review]

**Pull request closed without merging**